optuna/integration/wandb.py

```diff
@@ -0,0 +1,144 @@
+
+    Example:
+
+        Add Weights & Biases callback to Optuna optimization.
+
+        .. code::
+
+            import optuna
+            from optuna.integration.wandb import WeightsAndBiasesCallback
+
+
+            def objective(trial):
+                x = trial.suggest_float("x", -10, 10)
+                return (x - 2) ** 2
+
+
+            n_trials = 10
+            wandbc = WeightsAndBiasesCallback(
+                n_trials=n_trials,
+                project_name="my-project-name",
+                group_name="my-group-name",
+                job_type="my-job-type",
+            )
+
+
+            study = optuna.create_study(study_name="my_study")
+            study.optimize(objective, n_trials=n_trials, callbacks=[wandbc])
+
+    Args:
+        n_trials:
+            The number of optimizer trials.
+        metric_name:
+            Name of the optimized metric. Since the metric itself is just a number,
+            `metric_name` can be used to give it a name. So you know later
+            if it was roc-auc or accuracy.
+        project_name:
+            Name of the project under which run should be categorized.
+            If the project is not specified, the run is put in an "Uncategorized" project.
+        group_name:
+            Specifies a group under project into which run should be categorized.
+            Please refer to `Group Runs
+            <https://docs.wandb.ai/guides/track/advanced/grouping>`_ for more details.
+        entity_name:
+            Username or team name under which run should be logged. When set to `None`,
+            the run is logged under the current user.
+        run_name:
+            Specifies the display name in Weights & Biases UI for this run.
+            When set to `None`, random two word name is generated instead.
+        job_type:
+            Specifies type of the run, and is used as additional grouping
+            for the runs.
+    """
+
+    def __init__(
+        self,
+        n_trials: int,
+        metric_name: Union[str] = "value",
+        project_name: Optional[str] = None,
+        group_name: Optional[str] = None,
+        entity_name: Optional[str] = None,
+        run_name: Optional[str] = None,
+        job_type: Optional[str] = None,
+    ) -> None:
+
+        _imports.check()
+
+        self._n_trials = n_trials
+        self._metric_name = metric_name
+        self._project_name = project_name
+        self._group_name = group_name
+        self._entity_name = entity_name
+        self._run_name = run_name
+        self._job_type = job_type
+
+    def __call__(self, study: optuna.study.Study, trial: optuna.trial.FrozenTrial) -> None:
+
+        if trial.number == 0:
+            self._initialize_run(study, trial)
+
+        for key, value in trial.params.items():
+            wandb.log({key: value}, step=trial.number)
+
+        wandb.log({self._metric_name: trial.value}, step=trial.number)
+
+        # We need to know when study ends to allow
```

Comment:
How about focusing on single process optimisation at this PR? If so, I think we do not need to call the following block because wandb automatically finished the logging.

ref: https://docs.wandb.ai/ref/python/run#finish

Reply:
So if I understand you correctly, instead of logging entire Study as single run, we would log just a Trial in the same way `mlflow` integration does?

Reply:
Sorry for my unclear comment. I mean the condition `if trial.number + 1 == self._n_trials`  and `wandb.finish()` might be unnecessary for single process optimisation. I suppose the tracked results are the same as the current code.

Reply:
Sure, although it would be nice in the future to have methods like `on_optimize_start()` and `on_optimize_end()` available for the callbacks to run this kind of cleanup in. Could i propose this in different issue after this PR?

Reply:
Of course!